Src/IronPython/Modules/Builtin.cs

```diff
@@ -217,9 +217,15 @@ public static partial class Builtin {
         }
 
         [Documentation("")] // provided by first overload
-        public static object compile(CodeContext/*!*/ context, [BytesConversion]IList<byte> source, string filename, string mode, object flags = null, object dont_inherit = null, int optimize = -1)
-            // TODO: Detect encoding as prescribed in PEP 263
-            => compile(context, StringOps.RawDecode(context, source, "utf-8", "strict"), filename, mode, flags, dont_inherit, optimize);
+        public static object compile(CodeContext/*!*/ context, [BytesConversion]IList<byte> source, string filename, string mode, object flags = null, object dont_inherit = null, int optimize = -1) {
+            byte[] bytes = source as byte[] ?? ((source is Bytes b) ? b.GetUnsafeByteArray() : source.ToArray());
```

Comment:
I have some similar changes in https://github.com/IronLanguages/ironpython3/pull/599. Could you take a look and see if the two could be unified? I'm not sure if it's better one way or another...

Reply:
I see. Will look into it. I saw the title of your PR passing by but did not expect it to cross with my focus here, which is #22 ("good first issue" :smirk:). Since PEP-3120 builds on top of PEP-263, the latter one got dragged in.

Reply:
I was also surprised that it lead me there. Apparently importlib uses compile to load its Python modules.

Reply:
It is not easy to choose simply one way or another. I see some some good elements in #599 and I prefer some other solutions from #603. To summarize:

 1. We both agree that a `MemoryStream` backed content provider is needed in the base `IronPython` project. The implementations we chose are virtually the same. Since there was already a class like this in zipimport, I think it is better to reuse it my moving to the main project rather than duplicating the code. Also the name `MemoryStreamContentprovider` seems more accurate than `PythonStreamContentProvider`. This is what #603 does.

 2. I like the idea of factoring out parts of `compile` to `ValidateMode` (I would prefer calling it `ValidateCompileMode`) and `CompileHelper`.

 3. The conversion to `byte[]` in the `IList<Byte>` overload of `compile` in #599 is somewhat restrictive. The conversion in #603 is wider and follows the pattern from `StringOps.DoDecode`.

 4. #603 addresses all three: `compile`, `eval`, and `exec`, while #599 only `compile`. In #603, only `exec` goes through `SourceUnit`, using directly `MemoryStream` and `SourceCodeReader` in the other two cases (I did it so for performance reasons). #599 uses `SourceUnit` for `compile`, which makes the code more regular. I propose to adopt the #599 approach to `compile`, leaving `eval` as it is (unless some neat refactoring opportunity emerges).

 5. The the last change in `PythonContext` in #599 (reporting unknown encoding name, around line 994) is superseded by more in depth changes in #603. I think it is best to revert this particular change in #599 to prevent merge conflicts.

 6. #599 has some whitespace-only corrections in `PythonContext` that will generate merge conflicts with #603. Best to revert (some of) them.

Let me know what you think of it.

Reply:
Agreed on everything. Good analysis.

1. The reason I renamed it was because I wanted to use it with a `Stream` and so having `MemoryStream` in the name no longer seemed accurate. However, I think leaving it as `MemoryStreamContentProvider` in here is fine since I have not finalized #599 and may still change things.